lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -89,132 +93,142 @@ def on_leave_field(node, parent, visitor)
-              end
+        # A hook which is called whenever a field's max complexity is calculated.
+        # Override this method to capture individual field complexity details.
+        #
+        # @param scoped_type_complexity [ScopedTypeComplexity]
+        # @param defined_complexity [Numeric] Field's defined complexity only excluding child complexity
+        # @param max_complexity [Numeric] Field's maximum complexity including child complexity
+        # @param child_complexity [Numeric, nil] Field's child complexity
+        def field_complexity(scoped_type_complexity, defined_complexity:, max_complexity:, child_complexity: nil)
+        end
 
-              if scoped_children_for_key
-                child_complexity = merged_max_complexity_for_scopes(query, scoped_children_for_key)
-                # This is the _last_ one we visited; assume it's representative.
-                max_complexity = complexity_for_key.own_complexity(child_complexity)
-                complexity_for_keys[child_key] = max_complexity
+        # @param children_for_scope [Array<Hash>] An array of `scoped_children[scope]` hashes
+        # (`{field_key => complexity}`)
+        # @return [Integer] Complexity value for all these selections in the current scope
+        def merged_max_complexity(query, children_for_scope)
+          all_keys = []
+          children_for_scope.each do |c|
+            all_keys.concat(c.keys)
+          end
+          all_keys.uniq!
+          complexity_for_keys = {}
+
+          all_keys.each do |child_key|
+            scoped_children_for_key = nil
+            complexity_for_key = nil
+            children_for_scope.each do |children_hash|
+              next unless children_hash.key?(child_key)
+
+              complexity_for_key = children_hash[child_key]
+              if complexity_for_key.terminal?
+                # Assume that all terminals would return the same complexity
+                # Since it's a terminal, its child complexity is zero.
+                complexity = complexity_for_key.own_complexity(0)
+                complexity_for_keys[child_key] = complexity
+
+                field_complexity(complexity_for_key, defined_complexity: complexity, max_complexity: complexity, child_complexity: nil)
+              else
+                scoped_children_for_key ||= []
+                scoped_children_for_key << complexity_for_key.scoped_children
               end
             end
 
-            # Calculate the child complexity by summing the complexity of all selections
-            complexity_for_keys.each_value.inject(0, &:+)
+            next unless scoped_children_for_key
+
+            child_complexity = merged_max_complexity_for_scopes(query, scoped_children_for_key)
+            # This is the _last_ one we visited; assume it's representative.
+            max_complexity = complexity_for_key.own_complexity(child_complexity)
+            defined_complexity = complexity_for_key.own_complexity(0)
```

Comment:
It seems like this is evaluated in order to be passed to `field_complexity(...)`, is that right?

Since we don't need it, and we already pass the required inputs to `field_complexity(...)` how about we don't evaluate it here, and instead, let user-defined `field_complexity` hooks generate that value if they need it, eg 

```ruby 
def field_complexity(complexity_for_key, max_complexity:, child_complexity:)
  defined_complexity = complexity_for_key.own_complexity(0)
  # ... 
end 
```

That way, we don't do any throw-away work in the default case (and I don't get any tickets about how the complexity proc is unexpectedly called twice 😅 ). Would that work here?

Reply:
🎉 yes that was specific to our needs anyway so I'll remove it by default